cpp/src/arrow/compute/light_array.cc

```diff
@@ -80,27 +80,32 @@ KeyColumnArray KeyColumnArray::Slice(int64_t offset, int64_t length) const {
   KeyColumnArray sliced;
   sliced.metadata_ = metadata_;
   sliced.length_ = length;
-  uint32_t fixed_size =
-      !metadata_.is_fixed_length ? sizeof(uint32_t) : metadata_.fixed_length;
+  uint32_t fixed_size = metadata_.fixed_length;
 
   sliced.buffers_[0] =
       buffers_[0] ? buffers_[0] + (bit_offset_[0] + offset) / 8 : nullptr;
   sliced.mutable_buffers_[0] =
       mutable_buffers_[0] ? mutable_buffers_[0] + (bit_offset_[0] + offset) / 8 : nullptr;
   sliced.bit_offset_[0] = (bit_offset_[0] + offset) % 8;
 
-  if (fixed_size == 0 && !metadata_.is_null_type) {
+  if (metadata_.fixed_length == 0 && !metadata_.is_null_type) {
+    ARROW_DCHECK(is_bool_type() || is_other_fixed_width_types())
+        << "Expected BOOL type or OTHER FIXED WIDTH type but got a different type.";
     sliced.buffers_[1] =
         buffers_[1] ? buffers_[1] + (bit_offset_[1] + offset) / 8 : nullptr;
     sliced.mutable_buffers_[1] = mutable_buffers_[1]
                                      ? mutable_buffers_[1] + (bit_offset_[1] + offset) / 8
                                      : nullptr;
     sliced.bit_offset_[1] = (bit_offset_[1] + offset) % 8;
-  } else {
+  } else if (metadata_.fixed_length > 0) {
+    ARROW_DCHECK(is_dictionary_type() || is_binary_type() || is_large_binary_type())
```

Comment:
I am confused.  If the type is `int32` then it seems we would reach this branch.  `metadata_.fixed_length` would be `4` and `metadata_.is_null_type` would be `false` and `metadata._is_fixed_length` would be `true`.

Reply:
That's right. 

If you're worried about why variable and fixed lengths are handled with the same logic, then the answer is as follows.

As I understand it, Arrow maintains a separate buffer for storing values for variable lengths. 

Because of this, we only need to calculate the offset well for the address pointing to the value, so I don't think fixed or variable length is an important consideration when calculating the offset. This is why I didn't think it was a problem to calculate it with the same logic.

If I'm misunderstanding, I'd be grateful to hear your concerns, I'm still a newbie to Arrow.

Reply:
> If I'm misunderstanding, I'd be grateful to hear your concerns, I'm still a newbie to Arrow.

I do not think this is a misunderstanding.  The key column array marks variable data as `!is_fixed_length` and then `fixed_length` should store the width of the offset (4 for string/binary and 8 for large_string/large_binary)